### PR TITLE
Integrated GPFA

### PR DESCRIPTION
**Pull request by dizcza**
Only `gpfa` method is integrated here among all other choices in neural_trajectory.
We'll add new functionality once it's implemented by our researchers.

### CODE REVIEW

File: elephant/gpfa.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    ------
+    AssertionError
+        If `seqs` es empty.
+
+    """
+    assert len(seqs) > 0, "Got empty trials."
+
+    # Set cross-validation folds
+    num_trials = len(seqs)
+    # fdiv = np.linspace(0, num_trials, num_folds + 1).astype(np.int) if \
```
Comment (alperyeg) [inline elephant/gpfa.py]: Can be removed

File: elephant/gpfa.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    y_all = np.hstack(seqs_train['y'])
+    y_dim = y_all.shape[0]
+
+    if np.linalg.matrix_rank(np.cov(y_all)) < y_dim:
+        errmesg = 'Observation covariance matrix is rank deficient.\n' \
+                  'Possible causes: ' \
+                  'repeated units, not enough observations.'
+        raise ValueError(errmesg)
+
+    print('Number of training trials: {}'.format(len(seqs_train)))
```
Comment (alperyeg) [inline elephant/gpfa.py]: Maybe this should be optional. Either as logging, or as an optional verbose variable as it is done in the `gpfa_core.py`.

File: elephant/gpfa_src/gpfa_util.py [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+    Returns the solution to x b = a. Equivalent to MATLAB right matrix
+    division: a / b
+    """
+    return np.linalg.solve(b.T, a.T).T
+
+
+def logdet(A):
+    """
+    log(det(A)) where A is positive-definite.
+    This is faster and more stable than using log(det(A)).
```
Comment (alperyeg) [inline elephant/gpfa_src/gpfa_util.py]: Is this equivalent in python/numpy?

File: elephant/utils.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import quantities as pq
+
+
+def check_quantities(var, var_name):
```
Comment (alperyeg) [inline elephant/utils.py]: Is this module a general file to provide utilities for `elephant`? That should be indicated.

File: elephant/gpfa.py (hunk 2 above)
Comment (dizcza) [reply elephant/gpfa.py]: yeah, a logger is another thing I'm thinking on

**Comment by dizcza** [comment]
As a side note, gpfa can be implemented as a class. Doing so will reduce a large number of repeated arguments (like `verbose`, `num-of-iters`, etc.) that are passed all over the gpfa functionality. I guess, if we agree to this, it'll be a separate PR.

File: elephant/gpfa_src/gpfa_util.py (hunk 3 above)
Comment (dizcza) [reply elephant/gpfa_src/gpfa_util.py]: This is equivalent to `np.log(np.linalg.det(A))` and, what is more important, numerically stable. However, this function is actually _slower_ than numpy's version. Maybe because I've installed MKL.
A good catch. I'll add a special test.
Figure_1

File: elephant/test/test_gpfa.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+"""
+Unit tests for the neural_trajectory analysis.
```
Comment (Junji110) [inline elephant/test/test_gpfa.py]: neural_trajectory --> GPFA

File: elephant/test/test_gpfa.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+else:
+    HAVE_SKLEARN = True
+    from elephant.gpfa_src import gpfa_util
+    from elephant.gpfa import gpfa
+
+python_version_major = sys.version_info.major
+
+
+@unittest.skipUnless(HAVE_SKLEARN, 'requires sklearn')
+class NeuralTrajectoryTestCase(unittest.TestCase):
```
Comment (Junji110) [inline elephant/test/test_gpfa.py]: NeuralTrajectoryTestCase --> GPFATestCase

File: elephant/test/test_gpfa.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        # generate a positive definite matrix
+        matrix = np.random.randn(20, 20)
+        matrix = matrix.dot(matrix.T)
+        logdet_fast = gpfa_util.logdet(matrix)
+        logdet_ground_truth = np.log(np.linalg.det(matrix))
+        assert_array_almost_equal(logdet_fast, logdet_ground_truth)
+
+
+def suite():
+    suite = unittest.makeSuite(NeuralTrajectoryTestCase, 'test')
```
Comment (Junji110) [inline elephant/test/test_gpfa.py]: NeuralTrajectoryTestCase --> GPFATestCase

File: elephant/utils.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        input argument
+    var_name : str
+        variable name
+
+    Raises
+    -------
+    AssertionError
+        If the input `var` is not a `pq.Quantity`
+    """
+    assert isinstance(var, pq.Quantity), \
```
Comment (Junji110) [inline elephant/utils.py]: I don't see a merit in defining this function.
Does this increase readability/maintainability of code so much?

**Comment by Junji110** [comment]
I found there are number of cases where `assert` is used to check the type or shape of input variables, but I don't think this is a proper usage of `assert`; you should rather raise `ValueError` in such cases.
To my understanding, `assert` should be used to check the internal logic of the code, but not to check whether users use the code in a proper way.

File: elephant/gpfa.py [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+                  'repeated units, not enough observations.'
+        raise ValueError(errmesg)
+
+    if verbose:
+        print('Number of training trials: {}'.format(len(seqs_train)))
+        print('Number of test trials: {}'.format(len(seqs_test)))
+        print('Latent space dimensionality: {}'.format(x_dim))
+        print('Observation dimensionality: {}'.format(has_spikes_bool.sum()))
+
+    min_var_frac = 0.01
```
Comment (Junji110) [inline elephant/gpfa.py]: This hard-coded number should rather be given as a keyword-argument (with a default of 0.01) of the function

File: elephant/utils.py (hunk 8 above)
Comment (dizcza) [reply elephant/utils.py]: Yes for maintainability and consistency. It does not improve readability.
Reason: instead of typing the same thing each time in each function.

**Comment by dizcza** [comment]
> I found there are number of cases where `assert` is used to check the type or shape of input variables, but I don't think this is a proper usage of `assert`; you should rather raise `ValueError` in such cases.
> To my understanding, `assert` should be used to check the internal logic of the code, but not to check whether users use the code in a proper way.

Actually yes, `ValueError` better fits the philosophy of checking the function input, while asserts are used as "I swear this cannot happen". I'll go through input checking asserts and change them to ValueError.

File: elephant/utils.py (hunk 8 above)
Comment (dizcza) [reply elephant/utils.py]: Forgot to mention. I'll add this function check to all elephant's public functions.

Comment (Junji110) [reply elephant/utils.py]: Then you should define this function in that PR, not here.

**Comment by Junji110** [comment]
I've refactored unit tests and added a new test on dimensionality reduction performance of GPFA

File: elephant/gpfa.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+                GP noise variances
+            d: ndarray of shape (#units, 1)
+                observation mean
+            C: ndarray of shape (#units, #latent_vars)
+                mapping between the neuronal data space and the latent variable
+                space
+            R: ndarray of shape (#units, #latent_vars)
+                observation noise covariance
+
+    seqs_train: numpy.recarray
```
Comment (mdenker) [inline elephant/gpfa.py]: I am not sure I understand the advantage of using `recarray`  instead of a classic list of `dict` here? Would
```python
seqs_train['y'][0] # recarrays, trial 0
```
be preferrable over
```python
seqs_train[0]['y'] # list of dicts, trial 0
```
from a user perspective?

File: elephant/gpfa.py [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+
+
+def gpfa(data, bin_size=20*pq.ms, x_dim=3, em_max_iters=500):
+    """
+    Prepares data and calls functions for extracting neural trajectories.
+
+    Parameters
+    ----------
+
+    data : list
```
[the remark this answers is not in the document]
Comment (dizcza) [reply elephant/gpfa.py]: I also agree that the trial ID is useless here because it always goes as `range(len(data))`. I also added TODO below. Since this is GPFA-related issue only, we might change this behaviour now, not waiting for the next issue and corresponding fix (some people might start using it immediately). Discussion is needed.

File: elephant/gpfa.py [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+                neural data
+            * xsm: ndarray of shape (#latent_vars, #bins)
+                posterior mean of latent variables at each time bin
+            * Vsm: ndarray of shape (#latent_vars, #latent_vars, #bins)
+                posterior covariance between latent variables at each
+                timepoint
+            * VsmGP: ndarray of shape (#bins, #bins, #latent_vars)
+                posterior covariance over time for each latent variable
+
+    seqs_test: numpy.recarray
```
[the remark this answers is not in the document]
Comment (dizcza) [reply elephant/gpfa.py]: Since cross-validation is not implemented yet, we can remove `seqs_test` at all. Discussion is needed.

File: elephant/gpfa.py [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+neural trajectories. The parameters (C, d, R) are estimated from the data using
+FACTOR ANALYSIS technique. GPFA is simply a set of Factor Analyzers (FA),
+linked together in the low dimensional space by a Gaussian Process (GP).
+
+The analysis consists of the following steps:
+
+0) bin the data to get a sequence of N dimensional vectors for each time
+    bin, and choose the reduced dimension x_dim;
+
+1) run
```
[the remark this answers is not in the document]
Comment (dizcza) [reply elephant/gpfa.py]: Agree. Some smart words are needed.

File: elephant/gpfa.py [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+            * method: str, Method name.
+
+    Raises
+    ------
+    AssertionError
+        If `data` is an empty list.
+        If `bin_size` if not a `pq.Quantity`.
+        If `data[0][1][0]` is not a `neo.SpikeTrain`.
+
+    Examples
```
[the remark this answers is not in the document]
Comment (dizcza) [reply elephant/gpfa.py]: This is a numpy-doc reserved keyword. There is no "Example" for a singular one.

File: elephant/gpfa.py [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+    >>> from elephant.spike_train_generation import homogeneous_poisson_process
+    >>> data = []
+    >>> for trial in range(50):
+    >>>     n_channels = 20
+    >>>     firing_rates = np.random.randint(low=1, high=100,
+    >>>         size=n_channels) * pq.Hz
+    >>>     spike_times = [homogeneous_poisson_process(rate=rate)
+    >>>         for rate in firing_rates]
+    >>>     data.append((trial, spike_times))
+    >>> params_est, seqs_train, seqs_test, fit_info = gpfa(
```
[the remark this answers is not in the document]
Comment (dizcza) [reply elephant/gpfa.py]: We discussed it with Junji a bit. The most obvious "explanation" would be a plot. But it requires matplotlib. The docs example is shown only to show the usage code, not the usefulness. Discussion is needed.

File: requirements-extras.txt [hunk 16]
```diff
@@ -1,3 +1,4 @@
 pandas>=0.14.1
 scikit-learn
 mpi4py>=3.0.1
+tqdm
```
[the remark this answers is not in the document]
Comment (dizcza) [reply requirements-extras.txt]: Yes, I'll add tqdm in other functions as well. I already added some todos in the following PRs where tqdm should replace handwritten progress bar.

File: elephant/gpfa.py (hunk 10 above)
Comment (dizcza) [reply elephant/gpfa.py]: I fully agree not using `np.recarray`. The authors should provide a clear reason why they stick to np.recarray over python-friendly structures. Apart from confusing the users (and probably developers as well), it has several disadvantages during the test setup and future maintenance.

Comment (Junji110) [reply elephant/gpfa.py]: I also though to define `seqs_train` as a dict rather than a recarray, but I found that there are many part where `seqs_train['y']` is subjected to vector operation, which is not quickly done with a dict.
We could still define it as a dict of arrays (of the same length, i.e., # of trials) representing `y`, `x`, and so on, but it would require a total rewrite of the whole code.

By the way, for a recarray `seqs_train`, `seqs_train['y'][0]` is identical to `seqs_train[0]['y']`.

File: elephant/gpfa.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Junji110) [reply elephant/gpfa.py]: The trial ID is inherited from the original MATLAB code, and the idea behind this would be that trial IDs can start from arbitrary number instead of 0 or 1, and there could even be missing numbers in between.
I personally don't think this feature is necessary, and if we get rid of this, the input structure would be much simpler.

**Comment by dizcza** [comment]
TODO list:
1. Replace `np.recarray` -> `dict`: hide `np.recarray` from the user in GPFA internal private functional, but still operate with dicts as input and output type.
2. Add an example where test data is used as not a part of the training data.
3. Move functions to OOP.
